Add Microchip CAN-FD MCP251xFD drivers

talih0 opened the pull request:
Continues PR https://github.com/zephyrproject-rtos/zephyr/pull/31270

legoabram committed:
drivers: can: Implement MCP25xxFD driver

Implementation for Microchip MCP2517FD/MCP2518FD SPI based CAN-FD
controller.

Signed-off-by: Abram Early <abram.early@gmail.com>

talih0 committed:
drivers: can: mcp251xfd: Add driver

This continue PR #31270. The updated changes are:
- Updated to work with latest zephyr
- Inplace reads/writes of registers
- Batch read of RX messages when multiple messages can be read
- FIFO abstraction of RX/TEF queues
- Handle ivmif errors
- Use READ_CRC for register reads
- Use bitmasks instead of bitfield members
- Rename mcp25xxfd to mcp251xfd
- General cleanups

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 committed:
shields: Add CAN-FD MikroElektronika mcp2518fd click shield

Adds MikroElektronica click shield for the Microchip
MCP2518FD CAN-FD controller.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

henrikbrixandersen reviewed (changes requested):
Please split this PR into at least three separate PRs:
- One for the xmc47_relax_kit board changes.
- One for the adding the arduino_uno_click shield.
- Keep this one for adding the MCP25xxFD driver + the mcp2518fd shield.

talih0 commented:
> Please split this PR into at least three separate PRs:
> 
>     * One for the xmc47_relax_kit board changes.
> 
>     * One for the adding the arduino_uno_click shield.
> 
>     * Keep this one for adding the MCP25xxFD driver + the mcp2518fd shield.

sure, will do. I did it this way because sample/drivers/can/counter needs all commits to run on xmc47_relax_kit. I'll add the sample later after the separate PRs are merged.

talih0 commented:
On the mcp2518fd click shield, I changed the example to lpcxpresso55s28, so that's independent of my xmc47_relax_kit/arduino uno click shield commits.
The compilation for  lpcxpresso55s28 works,  but I don't have the board to test.

erwango commented on boards/shields/mcp2518fd_click/mcp2518fd_click.overlay:
```diff
@@ -0,0 +1,28 @@
+&mikrobus_spi {
+	cs-gpios = <&mikrobus_header 2 GPIO_ACTIVE_LOW>;
+	mcp2518fd: canfd@0 {
```
See https://docs.zephyrproject.org/latest/hardware/porting/shields.html#shield-porting-and-configuration:
> Besides, in order to avoid name conflicts with devices that may be defined at board level, it is advised, specifically for shields devicetree descriptions, to provide a device nodelabel is the form <device>_<shield>, ....

str4t0m reviewed (changes requested):
Thanks for this contribution @talih0.
I've done a 1st partial review. Mostly formatting and misc stuff. Will have a closer look later on. Could you please address these so far.

Did you run the test in /tests/drivers/can, and did they all pass?
It would be great if you could add an additional commit adding a configuration for the build_all test in `tests/drivers/build_all/can`, similarly as was done for the mcp2515 can controller.

ifyall commented:
@henrikbrixandersen , do you understand why zephyr-bot added the Infineon Platform label to this? Looking at the changes, I don't see an immediately obvious reason.

Ian

talih0 commented on boards/shields/mcp2518fd_click/doc/index.rst:
```diff
@@ -0,0 +1,39 @@
+Requirements
+************
+
+The shield uses a mikroBUS interface. The target board must define
+a mikrobus_spi node label (see :ref:`shields` for more details).
```
ah yes, good point, will update.

talih0 commented on boards/shields/mcp2518fd_click/mcp2518fd_click.overlay:
```diff
@@ -0,0 +1,28 @@
+		osc-freq = <40000000>;
+		bus-speed = <500000>;
+		bus-speed-data = <2000000>;
+		prop-seg = <0>;
+		phase-seg1 = <15>;
+		phase-seg2 = <4>;
+		sjw = <4>;
+		prop-seg-data = <0>;
+		phase-seg1-data = <7>;
+		phase-seg2-data = <2>;
+		sjw-data = <2>;
```
sounds good

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1550 @@
+#define MCP25XXFD_FIFO_UNLOCK(mutex)
+#endif
+
+static int mcp25xxfd_reset(const struct device *dev)
```
Configuration mode is the default at power on, but I'll do it explicitly in the function as well.

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1550 @@
+	memcpy(dst->DATA, src->data, MIN(can_dlc_to_bytes(src->dlc), CAN_MAX_DLEN));
+}
+
+void *mcp25xxfd_read_reg(const struct device *dev, uint16_t addr, int len)
+{
```
As I understand this issue only affects the READ_CRC command:
http://ww1.microchip.com/downloads/en/DeviceDoc/MCP2518FD-Silicon-Errata-Data-Sheet-Clarification-80000789A.pdf

But I do think it will be good to add READ_CRC in the future. I'll add a comment in the code.

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1550 @@
+	return ret;
+}
+
+static int mcp25xxfd_init_iocon_reg(const struct device *dev)
```
I couldn't find this note in the errata or documentation, is this from a microchip forum?

talih0 committed:
tests: drivers: build_all: can: Add mcp251xfd test case

Adds build test case for mcp25xxfd with arduino_uno_click and
mikroe_mcp2518fd_click shield.

Signed-off-by: Andriy Gelman <andriy.gelman@gmail.com>

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1550 @@
+	const struct spi_buf_set rx = {.buffers = &rx_buf, .count = 1};
+
+	ret = spi_transceive_dt(&dev_cfg->bus, &tx, &rx);
+	if (ret) {
```
I used (ret < 0) instead, because there were a few cases where ret > 0 is not an error case.

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1550 @@
+	int ret;
+
+	timing->sjw = timing_params->sjw;
+	if (timing_params->sample_point && USE_SP_ALGO) {
```
I fixed most examples, but I think in this case the output of the operator && is already boolean.

talih0 commented:
> Thanks for this contribution @talih0. I've done a 1st partial review. Mostly formatting and misc stuff. Will have a closer look later on. Could you please address these so far.
> 
> Did you run the test in /tests/drivers/can, and did they all pass? It would be great if you could add an additional commit adding a configuration for the build_all test in `tests/drivers/build_all/can`, similarly as was done for the mcp2515 can controller.

@str4t0m, I appreciate the thorough review and all the suggestions. Thank you!

Yes, all tests/drivers/build_all/can are now passing.

talih0 commented:
> @henrikbrixandersen , do you understand why zephyr-bot added the Infineon Platform label to this? Looking at the changes, I don't see an immediately obvious reason.
> 
> Ian

Hi @ifyall , I think it's because initially the arduino_r3_headers commit for xmc47_relax_kit was part of this PR

str4t0m reviewed (dismissed):
Thanks for addressing the comments so far. I have a couple more comments, mostly about the configuration.

str4t0m commented:
> Yes, all tests/drivers/build_all/can are now passing.

The build_all tests are mainly to make sure that the driver can be built without warning. Could you please check all the tests in tests/drivers/can as well?

talih0 commented on boards/shields/mikroe_mcp2518fd_click/doc/index.rst:
```diff
@@ -0,0 +1,41 @@
+Programming
+***********
+
+Set ``-DSHIELD=mcp2518fd_click`` when you invoke ``west build``,
```
thanks, I missed this one.

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1663 @@
+		goto done;
+	}
+
+	tdc = mcp25xxfd_get_spi_buf_ptr(dev, sizeof(*tdc));
+	if (is_nominal) {
+		tdc->TDCMOD = MCP25XXFD_TDCMOD_DISABLED;
+	} else {
+		tdc->TDCMOD = MCP25XXFD_TDCMOD_AUTO;
+		tdc->TDCO = timing->prescaler * (timing->prop_seg + timing->phase_seg1);
+	}
```
good idea, I'll update.

talih0 replied:
It seems that the loopback mode requires transmitter delay compensation to be disabled. Otherwise, the device goes into a bus-off state during FD messages.

This was unintentionally done during the tests because they contain several calls to can_set_timing() that disabled the TDCO. 
The tests would have actually failed if there were any extra calls to can_set_timing_data(). This is not done in the API test yet (at least can_set_timing_data() that is supposed to succeed).

I've fixed this issue in the latest version of the driver. The TDCO is enabled/disabled when changing the internal mode.

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1663 @@
+	return mcp25xxfd_write_reg(dev, MCP25XXFD_REG_TEFCON, sizeof(*fifocon));
+}
+
+static int mcp25xxfd_init_tx_fifos(const struct device *dev)
```
sounds good, I'll update.

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1663 @@
+	return spi_write_dt(&dev_cfg->bus, &tx);
+}
+
+static int mcp25xxfd_init(const struct device *dev)
```
we call mcp25xxfd_reset() in the init which will reset the filters, or did I misunderstand?

talih0 commented on drivers/can/can_mcp25xxfd.h:
```diff
@@ -0,0 +1,578 @@
+#define MCP25XXFD_RX_FIFO_IDX CONFIG_CAN_MCP25XXFD_MAX_TX_QUEUE
+#define MCP25XXFD_REG_SIZE    4
+
+BUILD_ASSERT(MCP25XXFD_RX_FIFO_ITEMS >= 1, "Cannot fit RX FIFO into RAM");
```
sounds good, I'll add it as an option

talih0 commented:
> > Yes, all tests/drivers/build_all/can are now passing.
> 
> The build_all tests are mainly to make sure that the driver can be built without warning. Could you please check all the tests in tests/drivers/can as well?

Yes, the api, shell, and timing tests are passing

talih0 commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1663 @@
+	struct mcp25xxfd_tscon *tscon = mcp25xxfd_get_spi_buf_ptr(dev, sizeof(*tscon));
+
+	tscon->TBCEN = 1;
+	tscon->TBCPRE = timing->prescaler - 1;
```
I made this a separate Kconfig option.

talih0 commented:
@str4t0m, I added the READ_CRC commands and use it for the register reads. Thanks for the suggestion.
The received messages are still fetched using the standard READ command at the moment because READ_CRC is only limited to fetch a max of 255 bytes (and as I understand from the errata, the RAM should not be affected after the messages are loaded).

An alternative could be to use READ_CRC if < 255 bytes are fetched and otherwise use the regular READ.

talih0 commented:
I need to rethink my lockless idea, because struct mcp25xxfd_spi_data also needs to be protected.

talih0 commented:
> I need to rethink my lockless idea, because struct mcp25xxfd_spi_data also needs to be protected.

I couldn't figure out a clean way to fix this. I decided to remove the lockless idea in the latest version.
I also made spi tx/rx buffers the same length so that the read/write reg can be done with a single dma transaction.

henrikbrixandersen commented on drivers/can/can_mcp25xxfd.h:
```diff
@@ -0,0 +1,605 @@
+#define MCP25XXFD_TDCMOD_MANUAL   0b01
+#define MCP25XXFD_TDCMOD_DISABLED 0b00
+
+#define MCP25XXFD_TIMING_MIN_INITIALIZER				\
+	{								\
+		.sjw = 1,						\
+		.prop_seg = 0,						\
+		.phase_seg1 = 2,					\
+		.phase_seg2 = 1,					\
+		.prescaler = 1,						\
+	}
+
+#define MCP25XXFD_TIMING_MAX_INITIALIZER				\
+	{								\
+		.sjw = 128,						\
+		.prop_seg = 0,						\
+		.phase_seg1 = 256,					\
+		.phase_seg2 = 128,					\
+		.prescaler = 256,					\
+	}
+
+#define MCP25XXFD_TIMING_DATA_MIN_INITIALIZER				\
+	{								\
+		.sjw = 1,						\
+		.prop_seg = 0,						\
+		.phase_seg1 = 1,					\
+		.phase_seg2 = 1,					\
+		.prescaler = 1,						\
+	}
+
+#define MCP25XXFD_TIMING_DATA_MAX_INITIALIZER				\
+	{								\
+		.sjw = 16,						\
+		.prop_seg = 0,						\
+		.phase_seg1 = 32,					\
+		.phase_seg2 = 16,					\
+		.prescaler = 256,					\
+	}
```
These are only used once. Please move them to where they are used instead of using macros. The macros used in other drivers are due to these having multiple front-ends. I don't see that as a use-case here?

talih0 replied:
ok, fair enough. I'll update.

henrikbrixandersen commented on drivers/can/can_mcp25xxfd.h:
```diff
@@ -0,0 +1,605 @@
+#define MCP25XXFD_REG_CON    0x000
+#define MCP25XXFD_REG_CON_B2 0x002
+#define MCP25XXFD_REG_CON_B3 0x003
+struct mcp25xxfd_con_b0 {
+	uint32_t DNCNT : 5;     /* Device Net Filter Bit Number */
+	uint32_t ISOCRCEN : 1;  /* Enable ISO CRC in CAN FD Frames */
+	uint32_t PXEDIS : 1;    /* Protocol Exception Event Detection Disabled */
+	uint32_t res0 : 1;
+} __packed;
```
Instead of using structs with upper-case bitfield members I would really prefer using bitmask macros, `FIELD_GET()`, and `FIELD_PREP()`. Packed bitfields in structs are not 100% portable and should be avoided.

This goes for all the register definitions in this file.

talih0 replied:
ok, I'll switch to bitfields.

henrikbrixandersen commented on drivers/can/can_mcp25xxfd.c:
```diff
@@ -0,0 +1,1792 @@
+	return ret;
+}
+
+static int mcp25xxfd_set_timing_nominal(const struct device *dev, const struct can_timing *timing)
+{
+	return mcp25xxfd_set_timing(dev, timing, true);
+}
+
+
+#if defined(CONFIG_CAN_FD_MODE)
+static int mcp25xxfd_set_timing_data(const struct device *dev, const struct can_timing *timing)
+{
+	return mcp25xxfd_set_timing(dev, timing, false);
+}
+#endif
```
Why not just get rid of `is_nominal` in `mcp25xxfd_set_timing()` and move the relevant code to each of these functions? That would make the code more readable.

talih0 replied:
I thought this way would remove code duplication. But I'll look over the part.

henrikbrixandersen reviewed (changes requested):
Thank you for taking over this work. Having support for the MCP25xxFD family would be a great addition.

I have added my initial review comments below. I have only skimmed the implementation file briefly, but I will review it again once the requested changes have been implemented.

talih0 commented on drivers/can/Kconfig.mcp25xxfd:
```diff
@@ -0,0 +1,76 @@
+	  higher priority. The thread is cooperative and will not be interrupted by
+	  another thread until execution is released.
+
+config CAN_MCP25XXFD_RX_TIMESTAMP_PRESCALER
+	int "Timestamp counter prescaler"
+	default 1
+	range 1 1024
+	depends on CAN_RX_TIMESTAMP
+	help
+	  The timestamp counter is derived from the internal clock divided by this value.
```
I'll add it to the deviceree.

talih0 commented on drivers/can/Kconfig.mcp25xxfd:
```diff
@@ -0,0 +1,76 @@
+
+if CAN_MCP25XXFD
+
+config CAN_MCP25XXFD_MODE_CHANGE_TIMEOUT_MSEC
+	int "Mode change timeout in milli seconds"
+	default 200
+	help
+	  Mode change timeout in milli seconds.
```
You are right, I don't think this should be configurable. I'll remove it.
(It just seemed like a convenient way to test different values.)

talih0 commented on drivers/can/Kconfig.mcp25xxfd:
```diff
@@ -0,0 +1,76 @@
+	default 5
+	range 1 31
+	help
+	  Defines the array size of the callback/msgq pointers.
+	  Must be at least the size of concurrent reads.
```
Will update.

erwango reviewed (dismissed):
Approved on shield side

henrikbrixandersen commented:
@talih0 Please rebase against `main` and ping me. I'd love to get this in before the v3.5 feature freeze next week.

talih0 commented:
@henrikbrixandersen, I pushed the rebased version.

alexanderwachter commented on dts/bindings/can/microchip,mcp251xfd.yaml:
```diff
@@ -0,0 +1,82 @@
+  sof-on-clko:
+    type: boolean
+    description: |
+      Output start-of-frame (SOF) signal on the CLKO pin each time every time
```
```suggestion
      Output start-of-frame (SOF) signal on the CLKO pin every time
```

talih0 replied:
thanks, I updated the PR.

jfischer-no reviewed (dismissed):
Shield LGTM

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+		return -ENOTSUP;
+	}
+
+	if (mode == CAN_MODE_NORMAL) {
+		dev_data->next_mcp251xfd_mode = MCP251XFD_REG_CON_MODE_CAN2_0;
+	}
+
+	if ((mode & CAN_MODE_FD) != 0) {
+#if defined(CONFIG_CAN_FD_MODE)
+		dev_data->next_mcp251xfd_mode = MCP251XFD_REG_CON_MODE_MIXED;
+#else
+		return -ENOTSUP;
+#endif
+	}
+
+	if ((mode & CAN_MODE_LISTENONLY) != 0) {
+		dev_data->next_mcp251xfd_mode = MCP251XFD_REG_CON_MODE_LISTENONLY;
+	}
+
+	if ((mode & CAN_MODE_LOOPBACK) != 0) {
+		dev_data->next_mcp251xfd_mode = MCP251XFD_REG_CON_MODE_EXT_LOOPBACK;
+	}
```
These lines will overwrite `dev_data->next_mcp251xfd_mode` - does this imply, that e.g. CAN-FD and external loopback cannot be enabled simultaneously?

talih0 replied:
Yes, that's right. Loopback is it's own separate mode:

![image](https://github.com/zephyrproject-rtos/zephyr/assets/17433307/c30e928a-7840-4b6a-8264-04101c61b6a5)

henrikbrixandersen replied:
So CAN-FD and loopback mode are mutually exclusive? Do the tests in `tests/drivers/can/api` pass?

talih0 replied:
Yes, all the tests pass. There are three that are skipped:

 - SKIP - [can_classic.test_recover] duration = 0.001 seconds  
 - SKIP - [can_classic.test_send_receive_ext_id_rtr] duration = 0.001 seconds                                                 
 - SKIP - [can_classic.test_send_receive_std_id_rtr] duration = 0.001 seconds

henrikbrixandersen replied:
And the CAN-FD tests pass as well?

talih0 replied:
Yes, all looks fine:
https://pastebin.com/a5SSSY5M

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+	}
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+	for (; mailbox_idx < MCP251XFD_TX_QUEUE_ITEMS; mailbox_idx++) {
```
Why not use

```suggestion
	for (mailbox_idx = 0; mailbox_idx < MCP251XFD_TX_QUEUE_ITEMS; mailbox_idx++) {
```

talih0 replied:
thanks, updated.

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+	__ASSERT(rx_cb != NULL, "rx_cb can not be null");
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
+
+	for (; filter_idx < CONFIG_CAN_MAX_FILTER ; filter_idx++) {
```
Same here?

talih0 replied:
thanks, updated.

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+	uint8_t *reg_byte;
+	uint32_t *reg;
+	int ret;
+
```
Needs a check against `CONFIG_CAN_MAX_FILTER` before using `filter_idx`?

talih0 replied:
thanks, updated.

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+#ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+static int mcp251xfd_recover(const struct device *dev, k_timeout_t timeout)
+{
+	struct mcp251xfd_data *dev_data = dev->data;
+
+	ARG_UNUSED(timeout);
+
+	if (!dev_data->started) {
+		return -ENETDOWN;
+	}
+
+	return -ENOTSUP;
```
Is this due to a hardware limitation?

talih0 replied:
That's right, it only has an auto-recovery mode.

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+
+	*reg = sys_le32_to_cpu(*reg);
+
+	if ((*reg & MCP251XFD_REG_TREC_TXBO) != 0) {
+		new_state = CAN_STATE_BUS_OFF;
+	} else if ((*reg & MCP251XFD_REG_TREC_TXBP) != 0) {
+		new_state = CAN_STATE_ERROR_PASSIVE;
+	} else if ((*reg & MCP251XFD_REG_TREC_RXBP) != 0) {
+		new_state = CAN_STATE_ERROR_PASSIVE;
+	} else if ((*reg & MCP251XFD_REG_TREC_RXWARN) != 0) {
+		new_state = CAN_STATE_ERROR_WARNING;
+	} else {
+		new_state = CAN_STATE_ERROR_ACTIVE;
+	}
+
+	if (new_state == dev_data->state) {
+		goto done;
+	}
+
+	tec = FIELD_GET(MCP251XFD_REG_TREC_TEC_MASK, *reg);
+	rec = FIELD_GET(MCP251XFD_REG_TREC_REC_MASK, *reg);
```
Instead of parsing these two in tow different functions, could `mcp251xfd_get_state()` be called here instead?

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+	return 0;
+}
+
+static const struct can_driver_api mcp251xfd_api_funcs = {
```
Please move these further down in the file similar to other driver implementations.

talih0 replied:
thanks, updated.

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(dev_cfg->int_gpio_dt.port)) {
```
```suggestion
	if (!gpio_is_ready_dt(dev_cfg->int_gpio_dt)) {
```

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+	gpio_init_callback(&(dev_data->int_gpio_cb), mcp251xfd_int_gpio_callback,
+			   BIT(dev_cfg->int_gpio_dt.pin));
+
+	if (gpio_add_callback(dev_cfg->int_gpio_dt.port, &(dev_data->int_gpio_cb)) < 0) {
```
```suggestion
	if (gpio_add_callback_dt(dev_cfg->int_gpio_dt, &(dev_data->int_gpio_cb)) < 0) {
```

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+
+	(void)k_thread_name_set(&dev_data->int_thread, "MCP251XFD interrupt thread");
+
+	k_mutex_lock(&dev_data->mutex, K_FOREVER);
```
Why is a lock needed in `init()`?

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+#endif
+
+#if defined(CONFIG_CAN_FD_MODE)
+	ret = can_set_mode(dev, CAN_MODE_FD);
```
All CAN controllers need to be in normal mode after boot up, regardless of `CONFIG_CAN_FD_MODE`.

henrikbrixandersen commented on drivers/can/can_mcp251xfd.c:
```diff
@@ -0,0 +1,1814 @@
+		MCP251XFD_SET_CLOCK(inst)                                                          \
+	};                                                                                         \
+                                                                                                   \
+	DEVICE_DT_INST_DEFINE(inst, &mcp251xfd_init, NULL, &mcp251xfd_data_##inst,                 \
```
```suggestion
	CAN_DEVICE_DT_INST_DEFINE(inst, &mcp251xfd_init, NULL, &mcp251xfd_data_##inst,                 \
```

talih0 commented on tests/drivers/build_all/can/testcase.yaml:
```diff
@@ -16,3 +16,8 @@ tests:
     platform_allow: frdm_k64f
+  drivers.build_all.can.mcp251xfd:
+    depends_on:
+      - arduino_spi
```
If we use extra_args: SHIELD=mikroe_mcp2518fd_click, then do we not need the arduino_spi dependency?
The tests above have the same code.

henrikbrixandersen replied:
Problem is, the dependency is not on `arduino_spi`, but on mikroe?

talih0 replied:
thanks, good point. :ooking at the lpcxpresso55s28 board it doesn't have mikrobus_spi in the supported list (even though mikrobus_spi is defined in the devicetree). Should I create a separate commit to add it to the list: 
![image](https://github.com/zephyrproject-rtos/zephyr/assets/17433307/e402c89a-553a-407d-be5d-d8dcaa157065)

henrikbrixandersen replied:
I think you can just remove the dependency for now. We can solve this in a follow-up PR.

talih0 replied:
sounds good, I removed the dependency in the latest push.

henrikbrixandersen commented:
@str4t0m, @erwango, @jfischer-no Please revisit.

kartben commented:
@jhedberg @fabiobaltieri in case there's still time to get it into RC1

talih0 commented:
thank you everyone for your reviews and getting this merged into 3.5

henrikbrixandersen commented:
> thank you everyone for your reviews and getting this merged into 3.5

Thank you for taking up this work and thank you to @legoabram for the initial PR. This driver is an important addition to the existing range of Zephyr CAN drivers.
